File: python/taflow/exponentially_weighted_sum.py
"""Persistent exponentially weighted moving sum."""

from typing import Any

import numpy as np

from ._native import ExponentiallyWeightedSumOperator as _Native
from ._series import as_float64_series


class ExponentiallyWeightedSum:
    """Compute a causal exponentially weighted sum using span ``timeperiod``.

    The smoothing factor is ``2 / (timeperiod + 1)`` and the recurrence is
    ``sum_t = x_t + (1 - alpha) * sum_(t-1)``.
    """

    def __init__(
        self,
        _input: Any,
        timeperiod: int = 14,
    ) -> None:
        self._state = _Native(timeperiod)
        self._length = 0
        self.extend(_input)

    def append(self, _input: float) -> "ExponentiallyWeightedSum":
        """Append one scalar observation."""
        self._state.append(float(_input))
        self._length += 1
        return self

    def extend(self, _input: Any) -> "ExponentiallyWeightedSum":
        """Append an aligned input series."""
        values = as_float64_series(_input)
        self._state.extend(values)
        self._length += len(values)
        return self

    def compute(self) -> np.ndarray:
        """Return the complete aligned weighted-sum history."""
        return self._state.compute()

    def __len__(self) -> int:
        """Return the number of observations consumed by this state."""
        return self._length

    @property
    def value(self) -> float | None:
        """Return the latest weighted sum."""
        return self._state.value

    def reset(self) -> "ExponentiallyWeightedSum":
        """Reset the recurrence and clear output history."""
        self._state.reset()
        self._length = 0
        return self

File: crates/taflow-core/src/indicators/exponentially_weighted_sum.rs
//! Causal exponentially weighted moving sum.

use crate::error::TaResult;
use crate::stream::ewm_alpha;

/// Persistent exponentially weighted sum with recurrence
/// `sum_t = x_t + (1 - alpha) * sum_(t-1)`.
#[derive(Debug, Clone)]
pub struct ExponentiallyWeightedSum {
    decay: f64,
    value: Option<f64>,
}

impl ExponentiallyWeightedSum {
    /// Create a state using `alpha = 2 / (timeperiod + 1)`.
    pub fn new(timeperiod: usize) -> TaResult<Self> {
        Ok(Self {
            decay: 1.0 - ewm_alpha(timeperiod)?,
            value: None,
        })
    }

    /// Append one observation and return the updated weighted sum.
    pub fn append(&mut self, input: f64) -> f64 {
        let value = input + self.decay * self.value.unwrap_or(0.0);
        self.value = Some(value);
        value
    }

    /// Return the latest weighted sum.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Clear accumulated weight while retaining the configured decay.
    pub fn reset(&mut self) {
        self.value = None;
    }
}
